From: Chad Monroe <chad@monroe.io>
Date: Tue, 09 Dec 2025 15:03:41 -0800
Subject: [PATCH] net: ethernet: mediatek: assign TX scheduler per GMAC

Assign each GMAC to a dedicated QDMA TX scheduler. Preserve scheduler
assignment when updating queue speed.

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c |   13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -825,7 +825,7 @@ static void mtk_set_queue_speed(struct m
 				int speed)
 {
 	const struct mtk_soc_data *soc = eth->soc;
-	u32 ofs, val;
+	u32 ofs, val, old_val;
 
 	if (!MTK_HAS_CAPS(soc->caps, MTK_QDMA))
 		return;
@@ -917,6 +917,13 @@ static void mtk_set_queue_speed(struct m
 	}
 
 	ofs = MTK_QTX_OFFSET * idx;
+	old_val = mtk_r32(eth, soc->reg_map->qdma.qtx_sch + ofs);
+
+	if (mtk_is_netsys_v2_or_greater(eth))
+		val |= old_val & MTK_QTX_SCH_TX_SEL_V2;
+	else
+		val |= old_val & MTK_QTX_SCH_TX_SEL;
+
 	mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
 }
 
@@ -2964,6 +2971,12 @@ static int mtk_tx_alloc(struct mtk_eth *
 			}
 			if (mtk_is_netsys_v1(eth))
 				val |= MTK_QTX_SCH_LEAKY_BUCKET_EN;
+
+			if (mtk_is_netsys_v2_or_greater(eth) && i == MTK_GMAC2_ID)
+				val |= FIELD_PREP(MTK_QTX_SCH_TX_SEL_V2, 1);
+			else if (mtk_is_netsys_v3_or_greater(eth) && i == MTK_GMAC3_ID)
+				val |= FIELD_PREP(MTK_QTX_SCH_TX_SEL_V2, 2);
+
 			mtk_w32(eth, val, soc->reg_map->qdma.qtx_sch + ofs);
 			ofs += MTK_QTX_OFFSET;
 		}
